src/backend/optimizer/path/allpaths.c

```diff
@@ -3134,6 +3134,21 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 		List *exprList = lappend(list_make1(subquery->havingQual),
 								 subpath->pathtarget->exprs);
 		path = turn_volatile_seggen_to_singleqe(root, path, (Node *) exprList);
+
+		/*
+		 * Sharing General and SegmentGeneral paths may lead to deadlock
+		 * when executed with 1-gang and joined with N-gang.
+		 */
+		if (is_shared && 
+			(CdbPathLocus_IsGeneral(path->locus) ||
+			 CdbPathLocus_IsSegmentGeneral(path->locus)))
+		{
+			Assert(IsA(path, CtePath));
+			CtePath *ctePath = (CtePath *) path;
+			ctePath->subpath = subpath;
+			cteplaninfo->subroot = NULL;
+		}
+
```

Comment:
1) Why is the solution much simpler than in 6X?
2) Why is the solution not the same as in 6X, because it would be possible to make changes after `subquery_planner`, as in 6X?
3) Maybe in 6X the solution can also be simplified like this?

Reply:
1. In 6X, we immediately create ShareInputScan and Material nodes. Therefore, 2 conditions are needed.
2. The sharing is disabled at this point because we do not want to disable it if the plan is executed on SingleQE due to the presence of volatile functions.
3. I don't think so.